### DISCUSSION TITLE
Add support for SKOS-XL

### DISCUSSION THREAD

**Discussion by koenedaele** [OnroerendErfgoed/atramhasis#926] [Ideas]
As requested by @mielvds 

Adding support for SKOS-XL allows making extra statements about a Label, such as adding a creation date or a specific type. Something like:

```Turtle
@prefix dcterms: <http://purl.org/dc/terms/> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
@prefix skosxl: <http://www.w3.org/2008/05/skos-xl#> .

<https://id.erfgoed.net/thesauri/erfgoedtypes/1005> a skos:Concept ;
    skosxl:prefLabel <https://id.erfgoed.net/thesauri/erfgoedtypes/1005/lbl1-nl> ;
    skos:scopeNote "Een kerk is een religieus gebouw waar christenen samenkomen om de liturgie/eucharistie te vieren. Een kerk onderscheidt zich soms enkel van een kapel door de grotere omvang."@nl-BE .

<https://id.erfgoed.net/thesauri/erfgoedtypes/1005/lbl1-nl> a skosxl:Label ;
    dcterms:created "2024-10-04" ;
    skosxl:literalForm "kerken"@nl .
```

### Observations

- Nothing similar exists for Notes. However, where the range skos:prefLabel is a Literal and the range of skosxl:prefLabel is a skosxl:Label, the range for skos:Note is undefined.
- Just adding support for SKOS-XL Labels does not bring a lot of extra functionality as it only adds a URI to a Label. This will only be interesting once we add extra data to such a Label and have users edit them. However, there are no such properties defined in SKOS-XL, so we will need to look further to other thesauri. There seems to be little overlap between eg. EUROVOC and AAT

### Examples

**EUROVOC**

```XML
<?xml version="1.0" encoding="utf-8" ?>
<rdf:RDF
	xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
	xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
	xmlns:dcterms="http://purl.org/dc/terms/"
	xmlns:owl="http://www.w3.org/2002/07/owl#"
	xmlns:skos="http://www.w3.org/2004/02/skos/core#"
	xmlns:skosxl="http://www.w3.org/2008/05/skos-xl#"
        xmlns:euvoc="http://publications.europa.eu/ontology/euvoc#" >
  <rdf:Description rdf:about="http://eurovoc.europa.eu/4101">
    <rdf:type rdf:resource="http://www.w3.org/2004/02/skos/core#Concept" />
    <skos:prefLabel xml:lang="nl">kerk</skos:prefLabel>
    <skosxl:prefLabel rdf:resource="http://eurovoc.europa.eu/319377" />
  </rdf:Description>
  <rdf:Description rdf:about="http://eurovoc.europa.eu/319377">
    <rdf:type rdf:resource="http://www.w3.org/2008/05/skos-xl#Label" />
    <rdf:type rdf:resource="http://eurovoc.europa.eu/schema#SimplePreferredTerm" />
    <dcterms:created rdf:datatype="http://www.w3.org/2001/XMLSchema#date">1995-10-02</dcterms:created>
    <owl:versionInfo>n/a</owl:versionInfo>
    <dcterms:type rdf:resource="http://publications.europa.eu/resource/authority/label-type/STANDARDLABEL" />
    <euvoc:status rdf:resource="http://publications.europa.eu/resource/authority/concept-status/CURRENT" />
    <skosxl:literalForm xml:lang="nl">kerk</ns4:literalForm>
  </rdf:Description>
</rdf:RDF>
```

**AAT**
```Turtle
<http://vocab.getty.edu/aat/300007466> a <http://vocab.getty.edu/ontology#Subject> , <http://www.w3.org/2004/02/skos/core#Concept> , <http://vocab.getty.edu/ontology#Concept> ;
    <http://www.w3.org/2004/02/skos/core#prefLabel> "kerken"@nl;
    <http://www.w3.org/2008/05/skos-xl#prefLabel> <http://vocab.getty.edu/aat/term/1000468080-nl> ;
    
<http://vocab.getty.edu/aat/term/1000468080-nl> a <http://www.w3.org/2008/05/skos-xl#Label> ;
	<http://vocab.getty.edu/ontology#term> "kerken"@nl ;
	<http://vocab.getty.edu/ontology#displayOrder> "18"^^<http://www.w3.org/2001/XMLSchema#positiveInteger> ;
	<http://www.w3.org/2008/05/skos-xl#literalForm> "kerken"@nl ;
	<http://vocab.getty.edu/ontology#termType> <http://vocab.getty.edu/term/type/Descriptor> ;
	<http://vocab.getty.edu/ontology#contributorPreferred> <http://vocab.getty.edu/aat/contrib/10000205> ;
	<http://purl.org/dc/terms/contributor> <http://vocab.getty.edu/aat/contrib/10000205> ;
	<http://vocab.getty.edu/ontology#sourcePreferred> <http://vocab.getty.edu/aat/source/2000113770> ;
	<http://purl.org/dc/terms/source> <http://vocab.getty.edu/aat/source/2000113770> ;
	<http://purl.org/dc/elements/1.1/identifier> "1000468080" ;
	<http://purl.org/dc/terms/language> <http://vocab.getty.edu/aat/300388256> , <http://vocab.getty.edu/language/nl> .
```

**LOC**
```XML
<rdf:RDF xmlns:skos="http://www.w3.org/2004/02/skos/core#" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#" xmlns:rdfs="http://www.w3.org/1999/02/22-rdf-schema#" xmlns:skosxl="http://www.w3.org/2008/05/skos-xl#">
  <rdf:Description rdf:about="http://id.loc.gov/authorities/subjects/sh85025502">
    <rdf:type rdf:resource="http://www.w3.org/2004/02/skos/core#Concept"/>
    <skos:prefLabel xml:lang="en">Church</skos:prefLabel>
    <skosxl:altLabel>
      <rdf:Description>
	<rdf:type rdf:resource="http://www.w3.org/2008/05/skos-xl#Label"/>
	<skosxl:literalForm xml:lang="en">Ecclesiastical theology</skosxl:literalForm>
      </rdf:Description>
    </skosxl:altLabel>
    <skosxl:altLabel>
      <rdf:Description>
	<rdf:type rdf:resource="http://www.w3.org/2008/05/skos-xl#Label"/>
	<skosxl:literalForm xml:lang="en">Ecclesiology</skosxl:literalForm>
      </rdf:Description>
    </skosxl:altLabel>
    <skosxl:altLabel>
      <rdf:Description>
	<rdf:type rdf:resource="http://www.w3.org/2008/05/skos-xl#Label"/>
	<skosxl:literalForm xml:lang="en">Theology, Ecclesiastical</skosxl:literalForm>
      </rdf:Description>
    </skosxl:altLabel>
    <skos:altLabel xml:lang="en">Ecclesiastical theology</skos:altLabel>
    <skos:altLabel xml:lang="en">Ecclesiology</skos:altLabel>
    <skos:altLabel xml:lang="en">Theology, Ecclesiastical</skos:altLabel>
  </rdf:Description>
</rdf:RDF>
```

LOC outputs some SKOS-XL but it adds not actual information.

From these examples it seems common to add a Type to a Label, however eurovoc adds types with both rdf:type and dcterms:type (from different vocabularies), while aat uses http://vocab.getty.edu/ontology#termType. The getty vocab types are Descriptor, AlternateDescriptor and UsedForTerm. Whereby it would seem to make sense that prefLabels are Descriptors and altLabels are AlternateDescriptors or UsedForTerm. However, http://vocab.getty.edu/aat/term/1000603616-de is a prefLabel for http://vocab.getty.edu/aat/300007466 and has type AlternateDescriptor. This doesn't seem very coherent.

### TODO

- Find more examples
- What use cases do we want to support in Atramhasis?

**Comment by mielvds** [comment]


**Reply by koenedaele** [reply]
I think these are all about versioning, not SKOS Labels, so I've moved them to #927

**Reply by mielvds** [reply]
oops, yes

**Comment by koenedaele** [comment]
Some information on specifics for SKOS labels: https://www.dublincore.org/specifications/skos-thes/ns/

**Reply by koenedaele** [reply]
skos-thes defines http://purl.org/iso25964/skos-thes#PreferredTerm and http://purl.org/iso25964/skos-thes#SimpleNonPreferredTerm as subclasses of skos-xl:Label and are the object of either skos-xl:prefLabel or skos-xl:altLabel/hiddenLabel. Although this adds no information to our model and is mainly useful when exporting data. skos-thes also supports Split Non Preferred Term, but I'm not very clear on the exact use case for this and why we would want to support this.

**Reply by koenedaele** [reply]
skos-thes also suggests using the following attributes on skos-xl:Label (https://www.dublincore.org/specifications/skos-thes/ns/#d4e643):

> Attributes or associations not detailed below typically are mapped to dc: (or dct:) properties:
> - dct:created
> - dct:modified
> - dc:source

**Comment by koenedaele** [comment]
SKOS-XL will be supported by Skosprovider in version 1.3.0. (https://github.com/OnroerendErfgoed/skosprovider/issues/125) Currently this doesn't really do a lot, except maintain a Label's URI in the Skosprovider environment. This will need to be handled further in Atramhasis by configurating generation of SKOS-XL label URI's.

**Comment by koenedaele** [comment]
Going back to this comment:

> From these examples it seems common to add a Type to a Label, however eurovoc adds types with both rdf:type and dcterms:type (from different vocabularies), while aat uses http://vocab.getty.edu/ontology#termType. The getty vocab types are Descriptor, AlternateDescriptor and UsedForTerm. 

The semantics between rdf:type and dcterms:type are quite different as rdf:type has a range of rdfs:Class, while dcterms:type has no range (but suggests using a controlled vocabulary). Finally, the getty:termType has a range of skos:Concept and seems closer to dcterms:type in meaning. On the one hand, we have rdf:type that changes the Label itself (possibly used to add rules about new properties), while dcterms:type and getty:termType are more about adding a simple type to the label.

The meemoo example also use dcterms:type:

```turtle
@prefix dct: <http://purl.org/dc/terms/> .
@prefix label-type: <http://publications.europa.eu/resource/authority/label-type/> .
@prefix ond: <https://w3id.org/onderwijs-vlaanderen/id/> .
@prefix ostr: <https://w3id.org/onderwijs-vlaanderen/id/structuur/> .
@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
@prefix skosxl: <http://www.w3.org/2008/05/skos-xl#> .

ostr:secundair-3e-graad-doorstroom-aso-short-label a skosxl:Label ;
    dct:type label-type:SHORTLABEL ;
    skosxl:literalForm "aso"@nl .

ostr:secundair-3e-graad-doorstroom-aso a skos:Concept ;
    skos:broader ostr:secundair-3e-graad-doorstroom ;
    skos:definition "secundair 3e graad finaliteit doorstroom aso"@nl ;
    skos:inScheme ond:structuur ;
    skos:prefLabel "secundair 3e graad finaliteit doorstroom aso"@nl ;
    skosxl:altLabel ostr:secundair-3e-graad-doorstroom-aso-short-label .
```

From these, I would make the following suggestions:
- We support dcterms:type in the skosprovider datamodel to add types to labels (but possibly also to concepts, collections or conceptschemes)
- Individual skosproviders (such as the getty vocab provider) could map this to more specific properties such as termType
- When outputting RDF in skosprovider_rdf, we use dcterms:type
- We currently do not support rdf:type

These first changes can go into skosprovider, making all of this editable in Atramhasis requires some decision about how to handle this in the UI and for data validation. Eg. do we want to configure possible types per conceptscheme? And where do we get the definitions for the labels from?

**Reply by mielvds** [reply]
> Going back to this comment:
> 
> > From these examples it seems common to add a Type to a Label, however eurovoc adds types with both rdf:type and dcterms:type (from different vocabularies), while aat uses http://vocab.getty.edu/ontology#termType. The getty vocab types are Descriptor, AlternateDescriptor and UsedForTerm.
> 
> The semantics between rdf:type and dcterms:type are quite different as rdf:type has a range of rdfs:Class, while dcterms:type has no range (but suggests using a controlled vocabulary). Finally, the getty:termType has a range of skos:Concept and seems closer to dcterms:type in meaning. 

Yes, `dcterms:type` has little to no semantics. Its mostly up to the user for interpretation. `rdf:type` does have semantics which will be interpreted by reasoners and rdf in general. That's not necessarily an issue. The choice really depends on what you need.

> On the one hand, we have rdf:type that changes the Label itself (possibly used to add rules about new properties), while dcterms:type and getty:termType are more about adding a simple type to the label.

Using rdf:type does not change the label itself. `<a> a skosxl:Label, :SomeClass` means that <a> is both a `skosxl:Label` and a `:SomeClass`, but that's all it says. It will not intervene in the interpretation of the data as SKOS .

> 
> The meemoo example also use dcterms:type:
> 
> ```turtle
> @prefix dct: <http://purl.org/dc/terms/> .
> @prefix label-type: <http://publications.europa.eu/resource/authority/label-type/> .
> @prefix ond: <https://w3id.org/onderwijs-vlaanderen/id/> .
> @prefix ostr: <https://w3id.org/onderwijs-vlaanderen/id/structuur/> .
> @prefix skos: <http://www.w3.org/2004/02/skos/core#> .
> @prefix skosxl: <http://www.w3.org/2008/05/skos-xl#> .
> 
> ostr:secundair-3e-graad-doorstroom-aso-short-label a skosxl:Label ;
>     dct:type label-type:SHORTLABEL ;
>     skosxl:literalForm "aso"@nl .
> 
> ostr:secundair-3e-graad-doorstroom-aso a skos:Concept ;
>     skos:broader ostr:secundair-3e-graad-doorstroom ;
>     skos:definition "secundair 3e graad finaliteit doorstroom aso"@nl ;
>     skos:inScheme ond:structuur ;
>     skos:prefLabel "secundair 3e graad finaliteit doorstroom aso"@nl ;
>     skosxl:altLabel ostr:secundair-3e-graad-doorstroom-aso-short-label .
> ```
> 

Yep, although I don't know why Ghent University - IDLab used dct:type here, but probably to avoid adding to formal type semantics. Maybe @mvanbrab or @bjdmeest could weigh in

> From these, I would make the following suggestions:
> 
>     * We support dcterms:type in the skosprovider datamodel to add types to labels (but possibly also to concepts, collections or conceptschemes)
> 
>     * Individual skosproviders (such as the getty vocab provider) could map this to more specific properties such as termType
> 
>     * When outputting RDF in skosprovider_rdf, we use dcterms:type
> 
>     * We currently do not support rdf:type

The choice to use `rdf:type` instead of `dcterms:type` is deliberate. We want the explicit semantics. We don't have this use case for labels, so using `dcterms:type` here is fine. 

However, for concepts, the use of `rdf:type` is a must, because we often use conceptschemes to organize class instances like these of `premis:StorageMedium`: https://github.com/viaacode/datamodels/blob/main/objects/carriers.skos.ttl or file formats: https://raw.githubusercontent.com/viaacode/datamodels/refs/heads/main/objects/formats.skos.ttl. This pattern is also used by Library of Congress: https://id.loc.gov/vocabulary/preservation/copyrightStatus/pub.html.
I suggest we discuss types for non-labels in a seperate thread?

We could also capture this in a query or rule, but that would require extra processing. However, I could live with it

> 
> 
> These first changes can go into skosprovider, making all of this editable in Atramhasis requires some decision about how to handle this in the UI and for data validation. Eg. do we want to configure possible types per conceptscheme? And where do we get the definitions for the labels from?

An elegant approach would be to use another (hidden) concept scheme (of types) to populate a type dropdown UI element. Or reuse the modal used in the 'relations' tab like here
<img width="463" height="519" alt="Screenshot 2025-09-24 at 11 56 31" src="https://github.com/user-attachments/assets/cfa45afb-dc10-4bdc-99a0-6fe116e8960c" />

**Reply by koenedaele** [reply]
Yes, how I envision it working is having the types come from a second concept scheme (could be internal or even external for something such as http://publications.europa.eu/resource/authority/label-type/). If we don't store the data in a concept scheme, we need to create something else for lookuplists that would basically amount to a SKOS vocabulary that's not formally a SKOS vocabulary and I don't see the point in that. This does imply that for in the skosprovider context `dct:type rdfs:range skos:Concept`, which I'm fine with. I'd apply this to the entire domain model, not just the labels.

>> On the one hand, we have rdf:type that changes the Label itself (possibly used to add rules about new properties), while dcterms:type and getty:termType are more about adding a simple type to the label.
>Using rdf:type does not change the label itself. `<a> a skosxl:Label, :SomeClass` means that <a> is both a `skosxl:Label` and a `:SomeClass`, but that's all it says. It will not intervene in the interpretation of the data as SKOS .

True, my way of expressing stuff isn't ideal. I meant something like "a Label with an `rdf:type` becomes more than a label or a label and something else", while a label with a `dct:type` is still just a `skos:Label`. Trying to understand what the different implementations could mean for an end-user.

**Reply by mielvds** [reply]
sure, but isn't the person attaching the `rdf:type` not aware of that? But again, for labels, I'm totally fine with using the safer `dct:type`. Or perhaps this can be configurable?

**Reply by bjdmeest** [reply]
> Yep, although I don't know why Ghent University - IDLab used dct:type here, but probably to avoid adding to formal type semantics. Maybe @mvanbrab or @bjdmeest could weigh in

This was mostly to follow the practice of how it's used in EuroVoc, e.g. see the rdf of https://op.europa.eu/en/web/eu-vocabularies/dataset/-/resource?uri=http://publications.europa.eu/resource/dataset/learning-activity, it contains things such as 

```turtle
<http://data.europa.eu/snb/learning-activity/label_b65952e63f33"> a skos-xl:Label ;
    dct:type <http://publications.europa.eu/resource/authority/label-type/STANDARDLABEL> .
```

**Reply by koenedaele** [reply]
I'm looking from the point of the Skosprovider interface, if it imports a SKOS file and at a later point exports it again, I want the meaning to be as similar as possible. So, semantically rdf:type is different from dct:type. Which seems to indicate we need to store them separately and make them editable in Atramhasis separatly as well. We could make it configurable, but Eurovoc used both of them side-by-side:

```rdf
 <rdf:Description rdf:about="http://eurovoc.europa.eu/4101">
    <rdf:type rdf:resource="http://www.w3.org/2004/02/skos/core#Concept" />
    <skos:prefLabel xml:lang="nl">kerk</skos:prefLabel>
    <skosxl:prefLabel rdf:resource="http://eurovoc.europa.eu/319377" />
  </rdf:Description>
 <rdf:Description rdf:about="http://eurovoc.europa.eu/319377">
    <rdf:type rdf:resource="http://www.w3.org/2008/05/skos-xl#Label" />
    <rdf:type rdf:resource="http://eurovoc.europa.eu/schema#SimplePreferredTerm" />
    <dcterms:created rdf:datatype="http://www.w3.org/2001/XMLSchema#date">1995-10-02</dcterms:created>
    <owl:versionInfo>n/a</owl:versionInfo>
    <dcterms:type rdf:resource="http://publications.europa.eu/resource/authority/label-type/STANDARDLABEL" />
    <euvoc:status rdf:resource="http://publications.europa.eu/resource/authority/concept-status/CURRENT" />
    <skosxl:literalForm xml:lang="nl">kerk</ns4:literalForm>
  </rdf:Description>
```

So, if on import we map all of them to dct:type we'll have lost some meaning. If we don't maintain the distinction, on export we will no longer know what came from rdf:type and what from dct:type. I do feel the example in question is bloated and I don't really understand where the need comes from to attacht that many types.

Anyway, the average user in Atramhasis has no idea whaat the difference between rdf:type and dct:type is and probably just wants to edit their concept. So, for the next implementation fase I'd like to add a simple type like dct:type to labels and that be fed with another concept scheme. That already solves the usecase of non-formal labeltypes.

**Reply by koenedaele** [reply]
See OnroerendErfgoed/skosprovider#146 for implementation in the Skosprovider library

**Reply by mielvds** [reply]
Looks we have an agreement then!
